File: app/src/main/java/com/example/homestay/ui/adapter/RoomAdapter.java
package com.example.homestay.ui.adapter;

import android.graphics.Color;
import android.view.*;
import android.widget.*;
import androidx.recyclerview.widget.RecyclerView;
import com.example.homestay.R;
import com.example.homestay.data.entity.Room;
import com.example.homestay.utils.ImageLoader;
import java.text.*;
import java.util.*;

public class RoomAdapter extends RecyclerView.Adapter<RoomAdapter.Holder> {
  public interface RoomClick {
    void onRoom(Room room);
  }

  public interface FavoriteClick {
    void onFavorite(Room room, boolean active);
  }

  public interface FavoriteStatus {
    boolean isFavorite(long roomId);
  }

  private final RoomClick roomClick;
  private final FavoriteClick favoriteClick;
  private final FavoriteStatus favoriteStatus;
  private List<Room> items = Collections.emptyList();

  public RoomAdapter(RoomClick c, FavoriteClick f, FavoriteStatus s) {
    roomClick = c;
    favoriteClick = f;
    favoriteStatus = s;
  }

  public void submitList(List<Room> v) {
    items = v;
    notifyDataSetChanged();
  }

  @Override
  public int getItemCount() {
    return items.size();
  }

  @Override
  public Holder onCreateViewHolder(ViewGroup p, int t) {
    return new Holder(LayoutInflater.from(p.getContext()).inflate(R.layout.item_room, p, false));
  }

  @Override
  public void onBindViewHolder(Holder h, int p) {
    h.bind(items.get(p));
  }

  public class Holder extends RecyclerView.ViewHolder {
    final ImageView image;
    final ImageButton favorite;
    final TextView name, location, rating, reviews, type, tag1, tag2, price, capacity;

    Holder(View v) {
      super(v);
      image = v.findViewById(R.id.img_room);
      favorite = v.findViewById(R.id.btn_favorite);
      name = v.findViewById(R.id.tv_room_name);
      location = v.findViewById(R.id.tv_location);
      rating = v.findViewById(R.id.tv_rating);
      reviews = v.findViewById(R.id.tv_review_summary);
      type = v.findViewById(R.id.tv_room_type);
      tag1 = v.findViewById(R.id.tv_amenities_tag);
      tag2 = v.findViewById(R.id.tv_amenities_tag_second);
      price = v.findViewById(R.id.tv_price);
      capacity = v.findViewById(R.id.tv_capacity);
    }

    void bind(Room r) {
      name.setText(r.getName());
      location.setText(r.getLocation());
      rating.setText(String.format(Locale.getDefault(), "%.1f", r.getRating()));
      reviews.setText(
          r.getReviewCount() == 0 ? "· Chưa có đánh giá" : "(" + r.getReviewCount() + ")");
      type.setText(r.getRoomType());
      type.setVisibility(r.getRoomType().isEmpty() ? View.GONE : View.VISIBLE);
      List<String> tags = tags(r.getAmenities());
      tag(tag1, tags.size() > 0 ? tags.get(0) : null);
      tag(tag2, tags.size() > 1 ? tags.get(1) : null);
      price.setText(
          NumberFormat.getNumberInstance(new Locale("vi", "VN")).format((long) r.getPrice())
              + " đ / đêm");
      capacity.setText("Tối đa " + r.getMaxGuests() + " khách · Còn phòng");
      int fallback = R.drawable.room_dalat;
      String data = r.getImageUrl().trim().isEmpty() ? null : r.getImageUrl();
      ImageLoader.load(image, data, fallback);
      boolean active = favoriteStatus != null && favoriteStatus.isFavorite(r.getId());
      setFavorite(active);
      favorite.setOnClickListener(
          v -> {
            boolean next = favoriteStatus == null || !favoriteStatus.isFavorite(r.getId());
            setFavorite(next);
            if (favoriteClick != null) favoriteClick.onFavorite(r, next);
          });
      itemView.setOnClickListener(v -> roomClick.onRoom(r));
    }

    void setFavorite(boolean active) {
      favorite.setImageResource(
          active ? android.R.drawable.btn_star_big_on : android.R.drawable.btn_star_big_off);
      favorite.setColorFilter(active ? Color.parseColor("#FFD700") : Color.WHITE);
    }

    void tag(TextView v, String s) {
      v.setText(s == null ? "" : s);
      v.setVisibility(s == null || s.isEmpty() ? View.GONE : View.VISIBLE);
    }
  }

  private static List<String> tags(String text) {
    List<String> out = new ArrayList<>();
    for (String raw : text.split(",")) {
      String s = raw.trim();
      if (!s.isEmpty()) {
        out.add(s.length() > 15 ? s.substring(0, 12) + "..." : s);
        if (out.size() == 2) break;
      }
    }
    return out;
  }
}

File: app/src/main/java/com/example/homestay/ui/adapter/RoomAdapter.kt
package com.example.homestay.ui.adapter

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageButton
import android.widget.ImageView
import android.widget.TextView
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import coil.load
import com.example.homestay.R
import com.example.homestay.data.entity.Room
import java.text.NumberFormat
import java.util.Locale

class RoomAdapter(
    private val onRoomClick: (Room) -> Unit,
    private val onFavoriteClick: ((Room, Boolean) -> Unit)? = null,
    private val getFavoriteStatus: ((Long) -> Boolean)? = null
) : ListAdapter<Room, RoomAdapter.RoomViewHolder>(RoomDiffCallback()) {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RoomViewHolder {
        val view = LayoutInflater.from(parent.context)
            .inflate(R.layout.item_room, parent, false)
        return RoomViewHolder(view)
    }

    override fun onBindViewHolder(holder: RoomViewHolder, position: Int) {
        holder.bind(getItem(position))
    }

    inner class RoomViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        private val imgRoom: ImageView = itemView.findViewById(R.id.img_room)
        private val btnFavorite: ImageButton = itemView.findViewById(R.id.btn_favorite)
        private val tvRoomName: TextView = itemView.findViewById(R.id.tv_room_name)
        private val tvLocation: TextView = itemView.findViewById(R.id.tv_location)
        private val tvRating: TextView = itemView.findViewById(R.id.tv_rating)
        private val tvRoomType: TextView = itemView.findViewById(R.id.tv_room_type)
        private val tvAmenitiesTag: TextView = itemView.findViewById(R.id.tv_amenities_tag)
        private val tvPrice: TextView = itemView.findViewById(R.id.tv_price)
        private val tvCapacity: TextView = itemView.findViewById(R.id.tv_capacity)

        fun bind(room: Room) {
            tvRoomName.text = room.name
            tvLocation.text = room.location
            // Format rating với 1 chữ số thập phân (5.0)
            tvRating.text = String.format(Locale.getDefault(), "%.1f (%d)", room.rating, room.reviewCount)
            
            // Room type - hiển thị nếu có
            if (room.roomType.isNotEmpty()) {
                tvRoomType.text = room.roomType
                tvRoomType.visibility = View.VISIBLE
            } else {
                tvRoomType.visibility = View.GONE
            }
            
            // Amenities tag - lấy amenity đầu tiên từ danh sách
            val amenitiesTag = getFirstAmenity(room.amenities)
            if (amenitiesTag.isNotEmpty()) {
                tvAmenitiesTag.text = amenitiesTag
                tvAmenitiesTag.visibility = View.VISIBLE
            } else {
                tvAmenitiesTag.visibility = View.GONE
            }
            
            // Format price
            val formatter = NumberFormat.getNumberInstance(Locale("vi", "VN"))
            val formattedPrice = formatter.format(room.price.toLong())
            tvPrice.text = "$formattedPrice đ / đêm"
            tvCapacity.text = "Tối đa ${room.maxGuests} khách · Còn phòng"

            // Load image from URL using Coil
            imgRoom.load(room.imageUrl) {
                placeholder(R.drawable.app_logo) // Placeholder while loading
                error(R.drawable.app_logo) // Error image if load fails
                crossfade(true) // Smooth transition
                listener(
                    onStart = { 
                        android.util.Log.d("RoomAdapter", "Loading image: ${room.imageUrl}")
                    },
                    onSuccess = { _, _ -> 
                        android.util.Log.d("RoomAdapter", "Image loaded successfully: ${room.imageUrl}")
                    },
                    onError = { _, result -> 
                        android.util.Log.e("RoomAdapter", "Failed to load image: ${room.imageUrl}, error: ${result.throwable.message}")
                    }
                )
            }

            // Setup favorite button
            val favorite = getFavoriteStatus?.invoke(room.id) ?: false
            updateFavoriteButton(favorite)

            btnFavorite.setOnClickListener {
                val newFavoriteState = !favorite
                updateFavoriteButton(newFavoriteState)
                onFavoriteClick?.invoke(room, newFavoriteState)
            }

            itemView.setOnClickListener {
                onRoomClick(room)
            }
        }

        private fun updateFavoriteButton(isFavorite: Boolean) {
            if (isFavorite) {
                btnFavorite.setImageResource(android.R.drawable.btn_star_big_on)
                btnFavorite.setColorFilter(android.graphics.Color.parseColor("#FFD700")) // Gold color
            } else {
                btnFavorite.setImageResource(android.R.drawable.btn_star_big_off)
                btnFavorite.setColorFilter(android.graphics.Color.WHITE)
            }
        }
        
        private fun getFirstAmenity(amenities: String): String {
            if (amenities.isEmpty()) return ""
            // Lấy amenity đầu tiên từ chuỗi (có thể là comma-separated hoặc JSON)
            val firstAmenity = amenities.split(",").firstOrNull()?.trim() ?: ""
            // Giới hạn độ dài để tránh tag quá dài
            return if (firstAmenity.length > 15) {
                firstAmenity.take(12) + "..."
            } else {
                firstAmenity
            }
        }
    }

    class RoomDiffCallback : DiffUtil.ItemCallback<Room>() {
        override fun areItemsTheSame(oldItem: Room, newItem: Room): Boolean {
            return oldItem.id == newItem.id
        }

        override fun areContentsTheSame(oldItem: Room, newItem: Room): Boolean {
            return oldItem == newItem
        }
    }
}

